File: day7_part2.py
import sys, os, time
from copy import deepcopy

class Day7Part2:
    DEBUG = False
    start_time = time.time()
    last_progress_time = start_time
    
    @staticmethod
    def parse_input(input_text):
        return input_text.splitlines()

    @staticmethod
    def find_item_indices(line, item): # item = beams (|) or splitters (^)
        items = []
        search_start = 0
        while True:
            n = line.find(item, search_start)
            if n >= 0:
                items.append(n)
                search_start = n + 1
            else:
                break
        return items
    @staticmethod
    def print_progress(count):
        now = time.time()
        if now - Day7Part2.last_progress_time > 2.0:
            rate = count / (now - Day7Part2.start_time)
            print(f"{count:,} - {rate:,.0f} / sec", end="\r") # progress indicator
            Day7Part2.last_progress_time = now
        
    # .......S.......
    # .......|.......
    # .......^.......
    # ...............
    # ......^.^......
    # ...............
    # .....^.^.^.....
    # ...............
    # ....^.^...^....
    # ...............
    # ...^.^...^.^...
    # ...............
    # ..^...^.....^..
    # ...............
    # .^.^.^.^.^...^.
    # ...............

    @staticmethod
    def iterate(lines, all_splitters, n, count, indent):
        Day7Part2.print_progress(count)
        if Day7Part2.DEBUG:
            print("\n")
            if n == 12 and count == 12:
                print('-'*100)
            print(f"{indent}n: {n}, count: {count}, len(lines): {len(lines)}")
        if n == len(lines):
            if Day7Part2.DEBUG:
                print(f"{indent}Hit end of input")
                for i in range(n):
                    print(f"{indent}{str(i).zfill(2)} {lines[i]}")
                print(f"{indent}1 Returning count {count}")
                print('-'*100)
            return count
        prev_beams = Day7Part2.find_item_indices(lines[n-1], '|')
        splitters = all_splitters[n] # Day7Part2.find_item_indices(lines[n], '^')
        if Day7Part2.DEBUG:
            for i in range(n-1):
                print(f"{indent}{str(i).zfill(2)} {lines[i]}")
            print(f"{indent}{str(n-1).zfill(2)} {lines[n-1]}, prev_beams: {prev_beams}")
            print(f"{indent}{str(n).zfill(2)} {lines[n]},  splitters: {splitters}")
        if len(prev_beams) != 1:
            print(f"Error: Beams should be 1. Was {len(prev_beams)}.")
        if splitters:
            no_splitter = True
            for splitter in splitters:
                if splitter in prev_beams:
                    no_splitter = False
                    # Optimization idea: Instead of using deepcopy, create left_lines with index 0 through n-1 as blank, then 
                    # copy n (modified) and the rest as is. I don't think I need to make a copy of all the [n+1:] lines, because
                    # they will get copied when I modify them later.
                    left_lines = ['.'] * n # We don't need the lines before row n
                    left_lines.append(lines[n][0:splitter-1] + '|' + lines[n][splitter:])
                    for line in lines[n+1:]:
                        left_lines.append(line) # Carry over the rest of the lines as-is
                    if Day7Part2.DEBUG:
                        print(f"{indent}Recursing left")
                    # Recursing left uses the same timeline, so we don't increase the count
                    count = Day7Part2.iterate(left_lines, all_splitters, n+1, count, indent+'  ') # propagate left
                    if Day7Part2.DEBUG:
                        print(f"{indent}Returned from recursing left, n: {n}")
                    
                    right_lines = ['.'] * n # We don't need the lines before row n
                    right_lines.append(lines[n][0:splitter+1] + '|' + lines[n][splitter+2:])
                    for line in lines[n+1:]:
                        right_lines.append(line) # Carry over the rest of the lines as-is
                    if Day7Part2.DEBUG:
                        print(f"{indent}Recursing right")
                    # Recursing right starts a new timeline, so we increase the count
                    count = Day7Part2.iterate(right_lines, all_splitters, n+1, count+1, indent+'  ') # propagate right
                    if Day7Part2.DEBUG:
                        print(f"{indent}Returned from recursing right, n: {n}")
            if no_splitter: # We didn't hit a splitter, so propagate down
                lines[n] = lines[n][0:prev_beams[0]] + '|' + lines[n][prev_beams[0]+1:]
                if Day7Part2.DEBUG:
                    print(f"{indent}1 Iterating down:")
                count = Day7Part2.iterate(lines, all_splitters, n+1, count, indent)
        else: # There were no splitters, so propagate down
            lines[n] = lines[n][0:prev_beams[0]] + '|' + lines[n][prev_beams[0]+1:]
            if Day7Part2.DEBUG:
                print(f"{indent}2 Iterating down:")
            count = Day7Part2.iterate(lines, all_splitters, n+1, count, indent)
        if Day7Part2.DEBUG:
            print(f"{indent}2 n:{n}, Returning count: {count}")
            print('-'*100)
        return count

    @staticmethod
    def iterate_tachyon_beam(lines):
        all_splitters = [Day7Part2.find_item_indices(line,'^') for line in lines]
        print(f"all_splitters: {all_splitters}") if Day7Part2.DEBUG else None
        start = lines[0].find('S')
        lines[1] = lines[1][:start] + '|' + lines[1][start+1:]
        timeline_count = Day7Part2.iterate(lines, all_splitters, 2, 1, '') # Start count at 1, since there's 1 timeline to start
        return timeline_count

if __name__ == "__main__":
    # If no filename is provided, try common input filenames before failing.
    if len(sys.argv) == 2:
        input_filename = sys.argv[1]
    else:
        # prefer Gayle's input if present, otherwise Dean's input
        if os.path.exists("day7_input_gayle.txt"):
            input_filename = "day7_input_gayle.txt"
        elif os.path.exists("day7_input_dean.txt"):
            input_filename = "day7_input_dean.txt"
        else:
            print(f"Usage: python {sys.argv[0]} <input_filename>")
            sys.exit(1)
  
    input_text = open(input_filename).read()
    lines = Day7Part2.parse_input(input_text)     
    result = Day7Part2.iterate_tachyon_beam(lines)
    print(f"Day 7 Part 2 result: {result}")

File: day7_part2.c
// C conversion of day7_part2.py.

// In Developer Powershell for VS 2019:
// cl day7_part2.c; .\day7_part2.exe day7_input_dean.txt

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// Full data is 142x142
#define MAX_LINES 150
#define MAX_LINE_LENGTH 150

// Globals for progress
int DEBUG = 0;
double start_time = 0;
double last_progress_time = 0;

void parse_input(const char* input_text, char lines[][MAX_LINE_LENGTH], int* num_lines);
void find_item_indices(const char* line, char item, int* items, int* num_items);
void print_progress(long count);
long iterate(int* all_beams, int** all_splitters, int n, long count, int num_lines);
long iterate_tachyon_beam(char lines[][MAX_LINE_LENGTH], int num_lines);

void parse_input(const char* input_text, char lines[][MAX_LINE_LENGTH], int* num_lines) {
    // Split input_text into lines, store in lines array, set num_lines
    int line_idx = 0, char_idx = 0;
    const char* p = input_text;
    while (*p) {
        if (*p == '\n' || *p == '\r') {
            if (char_idx > 0) {
                lines[line_idx][char_idx] = '\0';
                line_idx++;
                char_idx = 0;
            }
            // Handle \r\n or \n\r
            if ((*p == '\r' && *(p+1) == '\n') || (*p == '\n' && *(p+1) == '\r')) {
                p++;
            }
        } else {
            if (char_idx < MAX_LINE_LENGTH-1) {
                lines[line_idx][char_idx++] = *p;
            }
        }
        p++;
    }
    if (char_idx > 0) {
        lines[line_idx][char_idx] = '\0';
        line_idx++;
    }
    *num_lines = line_idx;
}

void find_item_indices(const char* line, char item, int* items, int* num_items) {
    // Find all indices of 'item' in 'line', store in 'items', set 'num_items'
    int idx = 0;
    int found = 0;
    while (line[idx] != '\0') {
        if (line[idx] == item) {
            items[found++] = idx;
        }
        idx++;
    }
    *num_items = found;
}

void print_progress(long count) {
    // Print progress every 2 seconds: count, elapsed time, and rate
    double now = (double)clock() / CLOCKS_PER_SEC;
    if (start_time == 0) {
        start_time = now;
        last_progress_time = now;
    }
    double elapsed = now - start_time;
    if (now - last_progress_time > 2.0) {
        double rate = (elapsed > 0) ? (count / elapsed) : 0.0;
        printf("%ld splits in %.0f secs: %.0f / sec\r", count, elapsed, rate);
        fflush(stdout);
        last_progress_time = now;
    }
}

long iterate(int* all_beams, int** all_splitters, int n, long count, int num_lines) {
    // Stub: recursive logic
    return 0;
}

long iterate_tachyon_beam(char lines[][MAX_LINE_LENGTH], int num_lines) {
    // Find all splitters for each line
    int* all_splitters[MAX_LINES];
    int num_splitters[MAX_LINES];
    for (int i = 0; i < num_lines; ++i) {
        all_splitters[i] = (int*)malloc(MAX_LINE_LENGTH * sizeof(int));
        find_item_indices(lines[i], '^', all_splitters[i], &num_splitters[i]);
        // Initialize unused elements to -1
        for (int j = num_splitters[i]; j < MAX_LINE_LENGTH; ++j) {
            all_splitters[i][j] = -1;
        }
    }

    // Find the start index 'S' in the first line
    int start = -1;
    for (int i = 0; lines[0][i] != '\0'; ++i) {
        if (lines[0][i] == 'S') {
            start = i;
            break;
        }
    }
    if (start == -1) {
        for (int i = 0; i < num_lines; ++i) free(all_splitters[i]);
        return 0;
    }

    // Prepare all_beams array
    int all_beams[MAX_LINES];
    all_beams[0] = 0;
    all_beams[1] = start;
    for (int i = 2; i < num_lines; ++i) all_beams[i] = 0;

    // Print all_beams
    // printf("all_beams: [");
    // for (int i = 0; i < num_lines; ++i) {
    //     printf("%d", all_beams[i]);
    //     if (i < num_lines - 1) printf(", ");
    // }
    // printf("]\n");

    // Print all_splitters
    // for (int i = 0; i < num_lines; ++i) {
    //     printf("all_splitters[%d]: [", i);
    //     for (int j = 0; j < num_splitters[i]; ++j) {
    //         printf("%d", all_splitters[i][j]);
    //         if (j + 1 < num_splitters[i]) printf(", ");
    //     }
    //     printf("]\n");
    // }

    long result = iterate(all_beams, all_splitters, 2, 1, num_lines);

    for (int i = 0; i < num_lines; ++i) free(all_splitters[i]);
    return result;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Usage: %s <input_filename>\n", argv[0]);
        return 1;
    }

    // Read the contents of the input file
    FILE* f = fopen(argv[1], "rb");
    if (!f) {
        printf("Could not open file: %s\n", argv[1]);
        return 1;
    }
    fseek(f, 0, SEEK_END);
    long fsize = ftell(f);
    fseek(f, 0, SEEK_SET);
    char* input_text = (char*)malloc(fsize + 1);
    if (!input_text) {
        printf("Memory allocation failed.\n");
        fclose(f);
        return 1;
    }
    fread(input_text, 1, fsize, f);
    input_text[fsize] = '\0';
    fclose(f);

    // Parse input into lines
    char lines[MAX_LINES][MAX_LINE_LENGTH];
    int num_lines = 0;
    parse_input(input_text, lines, &num_lines);
    printf("Read %d lines from %s\n", num_lines, argv[1]);

    // Call iterate_tachyon_beam and print the result
    long result = iterate_tachyon_beam(lines, num_lines);
    printf("\nDay 7 Part 2 result: %ld\n", result);

    free(input_text);
    return 0;
}
